### ISSUE TITLE
Concurrent download of solc compiler fails

### ISSUE THREAD

**Issue by robercano** [NomicFoundation/hardhat#4382]
### Version of Hardhat

2.14.0

### What happened?

We are using Lerna in our monorepo, and we have 2 packages with contracts. A GitHub action is used to compile the packages when changes are pushed to the repo. This uses Lerna to concurrently compile the 2 packages with contracts. Both attempt to download the same compiler at the same time, and sometimes it fails. It seems a concurrency issue probably due to opening the file from 2 different threads.

We have a workaround in which we have a third dummy package with a dummy contract for the same compiler version. We first run the compilation of the dummy package, thus triggering the download of the compiler. Then when we compile the other 2 packages the compiler is already available and it works.

However this is a nasty approach that also adds a new dependency to our repo: every time we need to add a new compiler version to a package, we need to remember to add the same version to the dummy package so it downloads the proper compiler.

This bug was related, but it was never followed up: https://github.com/NomicFoundation/hardhat/issues/1925

### Minimal reproduction steps

Run 2 Hardhat compilations in parallel for the same repo.

### Search terms

concurrent download lerna hardhat compiler

**Comment by fvictorio** [comment]
Hey @robercano, thanks for this. I believe it's a duplicate of https://github.com/NomicFoundation/hardhat/issues/3722 though. But your reproduction repo is better than mine, so I'm going to comment it in that issue.

**Comment by robercano** [comment]
Great, thanks for checking it out! I'll jump into the other issue to follow it up
